// @sumsub/websdk/dist/LegacyBuilder.js
import SnsWebSdk from './SnsWebSdk';
import { getTokenVersion } from './common';
var Builder = /** @class */ (function () {
    function Builder(baseUrl, flowName) {
        this.debugEnabled = false;
        this.options = {
            adaptIframeHeight: true,
            addViewportTag: true,
            enableScrollIntoView: true,
        };
        this.config = null;
        this.reusableConfig = null;
        this.accessToken = null;
        this.expirationHandler = null;
        this.baseUrl = baseUrl;
        this.flowName = flowName;
    }
    Builder.prototype.withAccessToken = function (accessToken, expirationHandler) {
        this.accessToken = accessToken;
        if (!expirationHandler || typeof expirationHandler !== 'function') {
            throw new Error('Invalid parameter, "expirationHandler" must be a function');
        }
        this.expirationHandler = expirationHandler;
        return this;
    };
    Builder.prototype.debug = function (debug) {
        this.debugEnabled = debug;
        return this;
    };
    Builder.prototype.withOptions = function (options) {
        var _a;
        if (options.hasOwnProperty('adaptIframeHeight')) {
            this.options.adaptIframeHeight = options.adaptIframeHeight;
        }
        if (options.hasOwnProperty('addViewportTag')) {
            this.options.addViewportTag = options.addViewportTag;
        }
        this.options.enableScrollIntoView = (_a = options.enableScrollIntoView) !== null && _a !== void 0 ? _a : true;
        return this;
    };
    Builder.prototype.withConf = function (config) {
        this.config = config;
        return this;
    };
    Builder.prototype.withReusableKycConf = function (reusableConfig) {
        this.reusableConfig = reusableConfig;
        return this;
    };
    Builder.prototype.build = function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
        if (!this.accessToken || !this.expirationHandler) {
            throw new Error('Configure access token end the expiration handler before');
        }
        var version = getTokenVersion(this.accessToken);
        return new SnsWebSdk(this.baseUrl, 
        // Config
        {
            version: version,
            theme: (_a = this.config) === null || _a === void 0 ? void 0 : _a.theme,
            customizationName: (_b = this.config) === null || _b === void 0 ? void 0 : _b.customizationName,
            translationName: (_c = this.config) === null || _c === void 0 ? void 0 : _c.translationName,
            accessToken: this.accessToken,
            flowName: this.flowName,
            lang: (_d = this.config) === null || _d === void 0 ? void 0 : _d.lang,
            email: (_e = this.config) === null || _e === void 0 ? void 0 : _e.email,
            phone: (_f = this.config) === null || _f === void 0 ? void 0 : _f.phone,
            country: (_g = this.config) === null || _g === void 0 ? void 0 : _g.country,
            uiConf: (_h = this.config) === null || _h === void 0 ? void 0 : _h.uiConf,
            i18n: (_j = this.config) === null || _j === void 0 ? void 0 : _j.i18n,
            documentsByCountries: (_k = this.config) === null || _k === void 0 ? void 0 : _k.documentsByCountries,
            documentDefinitions: (_l = this.config) === null || _l === void 0 ? void 0 : _l.documentDefinitions,
            autoSelectDocumentDefinitions: (_m = this.config) === null || _m === void 0 ? void 0 : _m.autoSelectDocumentDefinitions,
            singleStep: (_o = this.config) === null || _o === void 0 ? void 0 : _o.singleStep,
            reusableConfig: this.reusableConfig,
        }, 
        // Callbacks
        {
            expirationHandler: {
                legacy: true,
                handler: this.expirationHandler,
            },
            onMessage: (_p = this.config) === null || _p === void 0 ? void 0 : _p.onMessage,
            onError: (_q = this.config) === null || _q === void 0 ? void 0 : _q.onError,
        }, 
        // Options
        {
            adaptIframeHeight: this.options.adaptIframeHeight,
            addViewportTag: this.options.addViewportTag,
            enableScrollIntoView: this.options.enableScrollIntoView,
            debug: this.debugEnabled,
        });
    };
    return Builder;
}());
export default Builder;
